awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -518,72 +525,79 @@ def create_if_needed(self, existing_item, new_item, endpoint, handle_response=Tr
+        # This will exit from the module on its own
+        # If the method successfully updates an item and on_update param is defined,
         #   the on_update parameter will be called as a method pasing in this object and the json from the response
-        # If you pass handle_response=False it will return one of three things:
+        # This will return one of three things:
         #    1. None if the existing_item does not need to be updated
         #    2. The response from Tower from patching to the endpoint. It's up to you to process the response and exit from the module.
         #    3. An ItemNotDefined exception, if the existing_item does not exist
-        # Note: common error codes from the Tower API can cause the module to fail even if handle_response is set to False
-        if existing_item:
-            # If we have an item, we can see if it needs an update
-            try:
-                item_url = existing_item['url']
-                item_type = existing_item['type']
-                if item_type == 'user':
-                    item_name = existing_item['username']
-                else:
-                    item_name = existing_item['name']
-                item_id = existing_item['id']
-            except KeyError as ke:
-                self.fail_json(msg="Unable to process update of item due to missing data {0}".format(ke))
-
-            needs_update = False
-            for field in new_item:
-                existing_field = existing_item.get(field, None)
-                new_field = new_item.get(field, None)
-                # If the two items don't match and we are not comparing '' to None
-                if existing_field != new_field and not (existing_field in (None, '') and new_field == ''):
-                    # Something doesn't match so let's update it
-                    needs_update = True
-                    break
-
-            if needs_update:
-                response = self.patch_endpoint(item_url, **{'data': new_item})
-                if not handle_response:
-                    return response
-                elif response['status_code'] == 200:
-                    self.json_output['changed'] = True
-                    self.json_output['id'] = item_id
-                    if on_update is None:
-                        self.exit_json(**self.json_output)
-                    else:
-                        on_update(self, response['json'])
-                elif 'json' in response and '__all__' in response['json']:
-                    self.fail_json(msg=response['json']['__all__'])
-                else:
-                    self.fail_json(**{'msg': "Unable to update {0} {1}, see response".format(item_type, item_name), 'response': response})
+        # Note: common error codes from the Tower API can cause the module to fail
+        if not existing_item:
+            self.fail_json(msg="The exstiing item is not defined and thus cannot be updated")
```

Comment:
I don't see how it's reasonable for the user to ever get this message.

I think a better solution would be to throw an exception, where it currently checks `if existing_item`, add an `else` and throw an exception, just about any exception would work fine.